pkg/io/v2/providers/providers.go

```diff
@@ -0,0 +1,192 @@
+		case providerS3:
+			return getS3Bucket(ctx, credentials, bucket)
+		default:
+			return nil, fmt.Errorf("passing credentials is only supported for S3 and GCS not %s", storageProvider)
+		}
+	}
+
+	bkt, err := blob.OpenBucket(ctx, fmt.Sprintf("%s://%s", storageProvider, bucket))
+	if err != nil {
+		return nil, fmt.Errorf("error opening file bucket: %v", err)
+	}
+	return bkt, nil
+}
+
+// getGCSBucket opens a gocloud blob.Bucket based on given credentials in the Google
+// credential JSON format (see documentation of GetBucket for an example)
+func getGCSBucket(ctx context.Context, credentials []byte, bucketName string) (*blob.Bucket, error) {
+	googleCredentials, err := google.CredentialsFromJSON(ctx, credentials, storage.ScopeFullControl)
+	if err != nil {
+		return nil, fmt.Errorf("error getting Google credentials from JSON: %v", err)
+	}
+
+	client, err := gcp.NewHTTPClient(
+		gcp.DefaultTransport(),
+		gcp.TokenSource(googleCredentials.TokenSource))
+	if err != nil {
+		return nil, fmt.Errorf("error creating GCP Http Client: %v", err)
+	}
+
+	bkt, err := gcsblob.OpenBucket(ctx, client, bucketName, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error opening GCS bucket: %v", err)
+	}
+	return bkt, nil
+}
+
+// s3Credentials are credentials used to access S3 or an S3-compatible storage service
+// Endpoint is an optional property. Default is the AWS S3 endpoint. If set, the specified
+// endpoint will be used instead.
+type s3Credentials struct {
+	Region           string `json:"region"`
+	Endpoint         string `json:"endpoint"`
+	Insecure         bool   `json:"insecure"`
+	S3ForcePathStyle bool   `json:"s3_force_path_style"`
+	AccessKey        string `json:"access_key"`
+	SecretKey        string `json:"secret_key"`
+}
+
+// getS3Bucket opens a gocloud blob.Bucket based on given credentials in the format the
+// struct s3Credentials defines (see documentation of GetBucket for an example)
+func getS3Bucket(ctx context.Context, creds []byte, bucketName string) (*blob.Bucket, error) {
+	s3Credentials := &s3Credentials{}
+	if err := json.Unmarshal(creds, s3Credentials); err != nil {
+		return nil, fmt.Errorf("error getting S3 credentials from JSON: %v", err)
+	}
+
+	staticCredentials := credentials.NewStaticCredentials(s3Credentials.AccessKey, s3Credentials.SecretKey, "")
+
+	sess, err := session.NewSession(&aws.Config{
+		Credentials:      staticCredentials,
+		Endpoint:         aws.String(s3Credentials.Endpoint),
+		DisableSSL:       aws.Bool(s3Credentials.Insecure),
+		S3ForcePathStyle: aws.Bool(s3Credentials.S3ForcePathStyle),
+		Region:           aws.String(s3Credentials.Region),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error creating S3 Session: %v", err)
+	}
+
+	bkt, err := s3blob.OpenBucket(ctx, sess, bucketName, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error opening S3 bucket: %v", err)
+	}
+	return bkt, nil
+}
+
+// ParseStoragePath parses storagePath and returns the storageProvider, bucket and relativePath
```

Comment:
Why do we need bucket to be a first class abstraction?

Reply:
I guess you mean why I'm not directly opening the file instead of the bucket and therefore if I have to parse the path here this way to find out the bucket name

* I wasn't aware of that but looks like I don't need the bucket name for OpenBucket. I can also just e.g: `blob.OpenBucket(context.TODO(), "gs://k8s-conform-capi-openstack/pr-logs")`
* Currently we only open one blob.Bucket per bucket and cache it. Imho this isn't possible if we open a Bucket for every single file

I tested the following:
```
bkt, err := blob.OpenBucket(context.TODO(), "gs://k8s-conform-capi-openstack/README.md")
attr, err := bkt.Attributes(context.TODO(), "")
content, err := bkt.ReadAll(context.TODO(), "")
reader, err := bkt.NewReader(context.TODO(), "", nil)
``
Only the first line worked, everything after that returned an `object name is empty` error.

The following works:
```
bkt, err := blob.OpenBucket(context.TODO(), "gs://k8s-conform-capi-openstack")
attr, err := bkt.Attributes(context.TODO(), "README.md")
content, err := bkt.ReadAll(context.TODO(), "README.md")
reader, err := bkt.NewReader(context.TODO(), "README.md", nil)
```

So I assume we can exchange bucket with something like path which then is the "directory". But I would also have to remove the caching per-bucket.